fe/fe-core/src/main/java/com/starrocks/qe/Coordinator.java

```diff
@@ -664,6 +684,210 @@ public void exec() throws Exception {
         }
     }
 
+    /**
+     * Compute the topological order of the fragment tree.
+     *
+     * @return multiple fragment groups. Each group should be delivered sequentially,
+     * and fragments in a group can be delivered concurrently.
+     */
+    private List<List<PlanFragment>> computeTopologicalOrderFragments() {
+        Queue<PlanFragment> queue = Lists.newLinkedList();
+        Map<PlanFragment, Integer> inDegrees = Maps.newHashMap();
+
+        PlanFragment root = fragments.get(0);
+        inDegrees.put(root, 0);
+        queue.add(root);
+        // Compute in-degree of each fragment.
+        while (!queue.isEmpty()) {
+            PlanFragment fragment = queue.poll();
+            for (PlanFragment child : fragment.getChildren()) {
+                Integer v = inDegrees.get(child);
+                if (v != null) {
+                    inDegrees.put(child, v + 1);
+                } else {
+                    inDegrees.put(child, 1);
+                    queue.add(child);
+                }
+            }
+        }
+
+        if (fragments.size() != inDegrees.size()) {
+            for (PlanFragment fragment : fragments) {
+                if (!inDegrees.containsKey(fragment)) {
+                    LOG.warn("This fragment does not belong to the fragment tree: {}", fragment.getFragmentId());
+                }
+            }
+            throw new StarRocksPlannerException("Some fragments do not belong to the fragment tree",
+                    ErrorType.INTERNAL_ERROR);
+        }
+
+        // `queue` contains the fragments whose in-degree is zero.
+        queue.add(root);
```

Comment:
Add root two times?

Reply:
No, `queue` is used two times as two times BFS.
- The first BFS computes in-degrees of each node, 
- and the second BFS computes node group according to the topological order.

Here is the initial stage of the second BFS, the queue must be empty since the first BFS is finished. Let me change variable names and add comments.
```C++
 // The first BFS.
while (!queue.isEmpty()) {
}

// The second BFS, queue must be empty.
queue.add(root);
```

Reply:
OK， I see